public/app/features/variables/state/actions.ts

```diff
@@ -448,3 +463,71 @@ const getQueryWithVariables = (getState: () => StoreState): UrlQueryMap => {
 
   return queryParamsNew;
 };
+
+export const DASHBOARD_SLOW_VARIABLES_TIMEOUT = 2000;
+export const DASHBOARD_CANCEL_SLOW_VARIABLES_TIMEOUT = 10000;
+
+export const initVariablesTransaction = (
+  dashboardUid: string,
+  dashboard: DashboardModel,
+  variableSrv: VariableSrv
+): ThunkResult<void> => async (dispatch, getState) => {
+  // Detect slow loading / initializing variables
+  // This is in order to show loading indication for slow variables queries
+  setTimeout(() => {
+    if (getState().templating.transaction.status === TransactionStatus.Fetching) {
+      dispatch(dashboardSetInitPhaseToSlowVariables());
+    }
+  }, DASHBOARD_SLOW_VARIABLES_TIMEOUT);
+
+  // Detect very slow loading / initializing variables
+  // This is in order to enable cancel for very slow running queries
+  setTimeout(() => {
+    if (getState().templating.transaction.status === TransactionStatus.Fetching) {
+      dispatch(dashboardSetInitPhaseToCancelVariables());
+    }
+  }, DASHBOARD_CANCEL_SLOW_VARIABLES_TIMEOUT);
```

Comment:
nitpick: There is no cancellation of these timeouts so they could accidentally fire during the loading of another dashboard. But the impact should be minimal.  

To keep complexity down I think we do not need all these init slow phases. I think the isInitSlow that governs the whole dashboard loading should be enough. And we can use that to show cancel action for the whole dashboard loading. thoughts? Just trying to keep it simple as a first step. 

I do like the new variables init phase! makes it clear what is taking time, that is super good / useful.

Reply:
Or if you prefer this maybe we can have a separate slow state? Just a bit complex with two different solutions for a slow state with this using initPhase & we still have dashboard isInitSlow to show loading spinner

Reply:
Good catch with the cancellations, and sure I'll remove them and just have initSlow.